src/Disks/IO/ThreadPoolRemoteFSReader.cpp

```diff
@@ -56,38 +56,58 @@ namespace
     };
 }
 
-IAsynchronousReader::Result RemoteFSFileDescriptor::readInto(char * data, size_t size, size_t offset, size_t ignore)
-{
-    return reader.readInto(data, size, offset, ignore);
-}
-
-
 ThreadPoolRemoteFSReader::ThreadPoolRemoteFSReader(size_t pool_size, size_t queue_size_)
-    : pool(std::make_unique<ThreadPool>(CurrentMetrics::ThreadPoolRemoteFSReaderThreads, CurrentMetrics::ThreadPoolRemoteFSReaderThreadsActive, pool_size, pool_size, queue_size_))
+    : pool(std::make_unique<ThreadPool>(CurrentMetrics::ThreadPoolRemoteFSReaderThreads,
+                                        CurrentMetrics::ThreadPoolRemoteFSReaderThreadsActive,
+                                        pool_size, pool_size, queue_size_))
 {
 }
 
 
 std::future<IAsynchronousReader::Result> ThreadPoolRemoteFSReader::submit(Request request)
 {
     ProfileEventTimeIncrement<Microseconds> elapsed(ProfileEvents::ThreadpoolReaderSubmit);
-    return scheduleFromThreadPool<Result>([request]() -> Result
-    {
-        CurrentMetrics::Increment metric_increment{CurrentMetrics::RemoteRead};
-        auto * remote_fs_fd = assert_cast<RemoteFSFileDescriptor *>(request.descriptor.get());
+    return scheduleFromThreadPool<Result>([request, this]() -> Result { return execute(request); },
+                                          *pool,
+                                          "VFSRead",
+                                          request.priority);
+}
 
-        auto async_read_counters = remote_fs_fd->getReadCounters();
-        std::optional<AsyncReadIncrement> increment = async_read_counters ? std::optional<AsyncReadIncrement>(async_read_counters) : std::nullopt;
+IAsynchronousReader::Result ThreadPoolRemoteFSReader::execute(Request request)
+{
+    CurrentMetrics::Increment metric_increment{CurrentMetrics::RemoteRead};
 
-        auto watch = std::make_unique<Stopwatch>(CLOCK_REALTIME);
-        Result result = remote_fs_fd->readInto(request.buf, request.size, request.offset, request.ignore);
-        watch->stop();
+    auto * fd = assert_cast<RemoteFSFileDescriptor *>(request.descriptor.get());
+    auto & reader = fd->getReader();
 
-        ProfileEvents::increment(ProfileEvents::ThreadpoolReaderTaskMicroseconds, watch->elapsedMicroseconds());
-        ProfileEvents::increment(ProfileEvents::ThreadpoolReaderReadBytes, result.size);
+    auto read_counters = fd->getReadCounters();
+    std::optional<AsyncReadIncrement> increment = read_counters ? std::optional<AsyncReadIncrement>(read_counters) : std::nullopt;
+
+    auto watch = std::make_unique<Stopwatch>(CLOCK_REALTIME);
+
+    reader.set(request.buf, request.size);
```

Comment:
nitpick: The SeekableReadBuffer is not supposed to be shared between RemoteFSFileDescriptor and anyone else, right? Otherwise we'd need to undo this `set` afterwards, in case someone wants to use `reader` without external buffer later. Consider adding a comment in RemoteFSFileDescriptor that it assumes the `reader` is not used by anything else.

Reply:
> The SeekableReadBuffer is not supposed to be shared between RemoteFSFileDescriptor and anyone else, right?

for reading yes, but it is hold by `AsynchronousBoundedReadBuffer` which can do `seek()` with it (which is actually equal to `reset()`) or some getters like `getFileOffsetOfBufferEnd()`.